client/rest/src/main/java/org/elasticsearch/client/RestClient.java

```diff
@@ -554,75 +593,133 @@ private void setHeaders(HttpRequest httpRequest, Header[] requestHeaders) {
-                //last resort: if there are no good hosts to use, return a single dead one, the one that's closest to being retried
-                List<Map.Entry<HttpHost, DeadHostState>> sortedHosts = new ArrayList<>(blacklist.entrySet());
-                if (sortedHosts.size() > 0) {
-                    Collections.sort(sortedHosts, new Comparator<Map.Entry<HttpHost, DeadHostState>>() {
-                        @Override
-                        public int compare(Map.Entry<HttpHost, DeadHostState> o1, Map.Entry<HttpHost, DeadHostState> o2) {
-                            return o1.getValue().compareTo(o2.getValue());
-                        }
-                    });
-                    HttpHost deadHost = sortedHosts.get(0).getKey();
-                    logger.trace("resurrecting host [" + deadHost + "]");
-                    nextHosts = Collections.singleton(deadHost);
+            long nanosUntilRevival = now - deadness.getDeadUntilNanos();
+            if (nanosUntilRevival > 0) {
+                livingNodes.add(node);
+                continue;
+            }
+            deadNodes.add(new DeadNodeAndRevival(node, nanosUntilRevival));
+        }
+
+        if (false == livingNodes.isEmpty()) {
+            /*
+             * Normal state: there is at least one living node. Rotate the
+             * list so subsequent requests will prefer the nodes in a
+             * different order then run them through the NodeSelector so it
+             * can have its say in which nodes are ok and their ordering. If
+             * the selector is ok with any over the living nodes then use
+             * them for the request.
+             */
+            // TODO this is going to send more requests to nodes right *after* a node that the selector removes
+            Collections.rotate(livingNodes, lastNodeIndex.getAndIncrement());
+            List<Node> selectedLivingNodes = nodeSelector.select(livingNodes);
+            if (false == selectedLivingNodes.isEmpty()) {
+                return selectedLivingNodes;
+            }
+        }
+
+        /*
+         * Last resort: If there are no good nodes to use, either because
+         * the selector rejected all the living nodes or because there aren't
+         * any living ones. Either way, we want to revive a single dead node
+         * that the NodeSelectors are OK with. We do this by sorting the dead
+         * nodes by their revival time and passing them through the
+         * NodeSelector so it can have its say in which nodes are ok and their
+         * ordering. If the selector is ok with any of the nodes then use just
+         * the first one in the list because we only want to revive a single
+         * node.
+         */
+        if (false == deadNodes.isEmpty()) {
+            Collections.sort(deadNodes, new Comparator<DeadNodeAndRevival>() {
+                @Override
+                public int compare(DeadNodeAndRevival lhs, DeadNodeAndRevival rhs) {
```

Comment:
DeadHostState is already `Comparable`, can we rely on that here too?

Reply:
I've been wondering about combining the `Node` into `DeadHostState` which I think would make this simpler. I'm kind of weary to do it as part of this already large change though.

Reply:
I wouldn't want to introduced more changes in this PR, but I would like to restore some of the previous stuff around comparable that was recently introduced.